### PR TITLE
Add BPF_MAP_TYPE_LRU_HASH

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Added BPF_MAP_TYPE_LRU_HASH to partially address #343.

Implemented as parallel key history hash-table. Need to investigate switching to heap for better reaping performance.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -27,6 +27,13 @@ typedef struct _ebpf_core_object_map
     ebpf_program_type_t program_type;
 } ebpf_core_object_map_t;
 
+typedef struct _ebpf_core_lru_map
+{
+    ebpf_core_map_t core_map;
+    // TODO: Replace with lock-free heap.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:33]: file an issue to track this?

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -459,13 +466,52 @@ _create_object_hash_map(_In_ const ebpf_map_definition_in_memory_t* map_definiti
...
+        if (retval != EBPF_SUCCESS) {
+            ebpf_hash_table_destroy((ebpf_hash_table_t*)map->core_map.data);
+            ebpf_free(map);
+            map = NULL;
+        }
+    }
+    if (map)
+        return &map->core_map;
+    else
+        return NULL;
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:496]: Could be shortened to
```suggestion
    return (map) ? &map->core_map : NULL;
```

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -557,11 +664,14 @@ _update_hash_map_entry(
...
 
     if ((entry_count == map->ebpf_map_definition.max_entries) &&
-        (ebpf_hash_table_find((ebpf_hash_table_t*)map->data, key, &value) != EBPF_SUCCESS))
+        (ebpf_hash_table_find((ebpf_hash_table_t*)map->data, key, &value) != EBPF_SUCCESS) &&
+        !_reap_oldest_map_entry(map))
         result = EBPF_INVALID_ARGUMENT;
     else
         result = ebpf_hash_table_update((ebpf_hash_table_t*)map->data, key, data, hash_table_operation);
 
+    _update_key_history(map, key, false);
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:673]: shouldn't this only be done in an EBPF_SUCCESS case?   In the 669 path, it's already done inside the ebpf_hash_table_find(), and in the 671 path if that fails there should be nothing to update.

**Review by dthaler — APPROVED** [review]
Ok but see comments

**Review by dthaler — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**